lib/virt_feature_test_base.pm

```diff
@@ -159,6 +165,7 @@ sub post_fail_hook {
     virt_utils::stop_monitor_guest_console;
     #(caller(0))[3] can help pass calling subroutine name into called subroutine
     $self->junit_log_provision((caller(0))[3]) if get_var("VIRT_AUTOTEST");
+    collect_virt_system_logs();
```

Comment:
@pdostal Would you please share the reasons why you finally choose to implement collect_virt_system_logs, rather than calling virt_utils::collect_host_and_guest_logs to upload logs? Actually the function can upload those logs. @waynechen55 @guoxuguang  FYI.

Reply:
Because of we load also tests such as:
 * `tests/console/perf.pm`
 * `tests/console/oprofile.pm`
 
 And as `virt_utils.pm` is located under `tests/virt_autotest/` there is no proper way to refer to it.
As we already spoke, this is by design broken, and I would like to avoid it as much as possible.

The next reason is all this hidden logic in bash, f.e.:
 * `virt_autotest/virt_logs_collector.sh`
 * `virt_autotest/fetch_logs_from_guest.sh`
I consider those as obsolete.

Reply:
> Because of we load also tests such as:
> 
> * `tests/console/perf.pm`
> * `tests/console/oprofile.pm`
> 
> And as `virt_utils.pm` is located under `tests/virt_autotest/` there is no proper way to refer to it.
> As we already spoke, this is by design broken, and I would like to avoid it as much as possible.
> 

@pdostal Thanks for explainations. I ask here because I hope all relevant people(both product and maintenance VT QA) can see from your reply why one side would like to abandon an existing function delivered by the other side and create a new one, so that in future we can avoid such abandon by better code design which considers both side's needs more to pursue higher code reuse. This will be beneficial to our squad.

> The next reason is all this hidden logic in bash, f.e.:
> 
> * `virt_autotest/virt_logs_collector.sh`
> * `virt_autotest/fetch_logs_from_guest.sh`
>   I consider those as obsolete.

No, they are not obsolete. They are totally new code to support VT logs collection, which collects more comprehensive VT logs, and survives the different conditions of ssh availability/no ssh, only serial console availability/only guest disk image.

Reply:
> > Because of we load also tests such as:
> 
> > * `tests/console/perf.pm`
> 
> > * `tests/console/oprofile.pm`
> 
> > And as `virt_utils.pm` is located under `tests/virt_autotest/` there is no proper way to refer to it.
> 
> > As we already spoke, this is by design broken, and I would like to avoid it as much as possible.
> 
> @pdostal Thanks for explainations. I ask here because I hope all relevant people(both product and maintenance VT QA) can see from your reply why one side would like to abandon an existing function delivered by the other side and create a new one, so that in future we can avoid such abandon by better code design which considers both side's needs more to pursue higher code reuse. This will be beneficial to our squad.

What we can do is that we can move the logic from to old function to the new one.


> > The next reason is all this hidden logic in bash, f.e.:
> 
> > * `virt_autotest/virt_logs_collector.sh`
> 
> > * `virt_autotest/fetch_logs_from_guest.sh`
> 
> >   I consider those as obsolete.
> 
> No, they are not obsolete. They are totally new code to support VT logs collection, which collects more comprehensive VT logs, and survives the different conditions of ssh availability/no ssh, only serial console availability/only guest disk image.

I am sorry, I tried to run those and I got empty files. When I tried to debug this, I discovered bash scripts which I was not able to orient in. My assumptions is that our guests were not configured as yours.

@guoxuguang Do you please have specific suggestions of what should I possibly improve?

Reply:
@pdostal 
1.Your collect_virt_system_logs just collects virt logs by copying some folders from host and additionally, guest xml config file. So it is quite simple but can server no other purposes because there are no room to make it more flexible from what I have seen till now. So if future requirements coming in that demands other types of logs, you have to insert more “if logic” to expand or invent another block to fulfill the specific requirements. I also admit that, under some circumstances, inventing new block to meet new requirements is inevitable, but I think we do not want it to go sideways in which monolithic way of development and difficult to maintain and extend can happen, because there will be too many branches, too many duplicates and too many impromptu additions due to maybe looming deadline or urgent usage.

2.Actually supportconfig can help collect nearly as many things as you want. It also provides many options to help customize collected logs. You should try it and see what you can get from it. And it is just one command to be fired up, so I can not see a reason that you do not use it. You may have your reasons but I have to say this single command help nearly do the same thing as your own function can. We also employ supportconfig in virt_utils::collect_host_and_guest_logs which demonstrate its effectiveness. And virt_utils::collect_host_and_guest_logs also provide additional interface to let you collect extra logs from host as well, you just need to give the the location which contains your extra logs.  As far as vm guests are concerned, supportconfig and extra logs can also be collected from them as well, either from ssh or guest console(if ssh is not available). So host and guests nearly share the same set of code block to collect logs. So virt_utils::collect_host_and_guest_logs looks like one-stop-logs-collecting tool.

3.Current virt_utils::collect_host_and_guest_logs collects vm guests logs from guest by using their ip addresses which are obtained from dhcpd lease file or guest console if ssh is not available. If can not get anything from guests by using virt_utils::collect_host_and_guest_logs, the reason may be 1) guest in shutdown state 2) can not get IP address because there is nothing in dhcpd lease file or dhcpd lease does not exist at all. This may be caused by different test environment. But you can implement your way of getting IP address  or even using guest domain name if you want. I think virt_utils::collect_host_and_guest_logs can be extended or tweaked to make it more all-round to work with this situation. Small modification can serve great purpose which deserved the effort. 3) Although connecting by ssh fails, guest console is the last resort. This will be another story if guest console can not be opened up. Maybe your guests do not have text console configured. Actually you can even totally skip guest log collecting if you do not want them by passing “none” as argument to virt_utils::collect_host_and_guest_logs

4.Summarizing up all my above arguments turns out to be: virt_utils::collect_host_and_guest_logs is better than collect_virt_system_logs in many aspects. And virt_utils::collect_host_and_guest_logs can be further extended to be more all-round. collect_virt_system_logs looks more like a quick and not-well-thought answer to a grand change(Judging by how many file you are going to modify), so they do not look match. I can accept collect_virt_system_logs as your final solution if this function is already very close to your final usage. But if the opposite is the case, I think you may end up modifying this function very frequently with continuously increasing new demands. Maybe you can share your thought further here to let me know what you are going to do, then the final answer may show up by itself. You may also have difficulty in using virt_utils::collect_host_and_guest_logs ? Maybe I can help :)

Reply:
Hello @waynechen55,

first of all, we cannot use `virt_utils::collect_host_and_guest_logs` as it isn't even includable.

1. We don't collect guest logs yet but my function is easy to ready and modify.
2. Supportconfig is actually a good idea! I will create poo# and implement it later.
3. The standard way is to use `virt_autotest::common` with static IP / MAC addresses.
4. Majority of changed files is `virt_feature_test_base` and ssh serial terminal implementation.

Thank you for your imput Wayne! ;)

Reply:
1. So how are you going to collect guest logs if you want ? By using ssh based on static ip from virt_autotest::common ?  And if ssh is not available, what are you going to do ?  I know this may be beyond the topic of this pull request, but I think your explanation may still help. I think even static "ip/mac" can be made available to virt_utils::collect_host_and_guest_logs, so you can use it for logs collecting for your guests as well. It may be "not includable" at the time, but it can be further tuned to be suitable to your usage as well. I hope you can see it as a basic framework or similar thing instead of something that is completely irrelevant. You said your function is easy to read and modify. Yes, it is and it has to be, because it does not include too many things in it. And "easy to modify" might be true if you think more and more  "if else" branches is acceptable. But it is kind of monolithic instead of modular, so it can be modified easily if new demand is also easy. It is not that good. So I suggest you can try virt_utils::collect_host_and_guest_logs and can make improvement suggestions and can even help contribute.
2. Actually logging thing is quite common one, especially in the same area. So it would be better to have a common framework or function or script or api to collect these for all. It will not only help decrease man effort, but also make automation development faster. So people can focus on the core test. It will benefit in the long run. So back to ITEM 1), you should think about it.
3. You did modify lib/virt_feature_test_base.pm a lot. I will have a quick look again and see whether there is anything unfit. To be honest, at least, collect_virt_system_logs seems not a very good solution for long-term use here. I will come back later to see if I can come up with something new  and any updates.

Reply:
1. I'm not planning to collect guest logs atm. And I don't think that modifying that new function would require many 'if else' branches.
2. I agree with this, but 1) the old function is not includable and 2) It contains a lot of hidden code in those bash scripts. That means it's not usable at all for us. Sorry.
3. I didn't modify `virt_feature_test_base` but I started to use it as a base class in many test modules.

Reply:
1. Could you help explain more about "not includable". So I think "not includable" leads to "not usable" right ?  Because I do not quite understand "not includable". I think every script and function in os-autoinst-distri-opensuse is "includable"
2.Yes, it is "hidden" because all those codes are not dispalyed in PM module directly. It is only hidden in this aspect.  But I do not think this prevents your from using or say you may choose to not use it but this should not be the reason. All functions in testapi.pm are hidden as well, but you call them every day.
3. You call your collect_virt_system_logs in virt_feature_test_base. And you started using it as base for your tests, so I think you may reply on collect_virt_system_logs more and more. There will be more and more duplicates between collect_virt_system_logs and virt_utils::collect_host_and_guest_logs. So please help explain more with regard to ITEM 1) and ITEM 2), we will know more about your situations. We may find your obstacles, and we may can help solve this together. I know you do not have any ongoing plan on logging, but if you can switch to virt_utils::collect_host_and_guest_logs, it can still bring benefits, for example, a) one universal virt_feature_test_base base class to some extent b) save your future effort on logging if any c) Just remove duplicates. 
I am not forcing you to switch. I just can see potential universal virt_utils::collect_host_and_guest_logs for both you and me, so I am asking for your input. Things are always building up, so is this case.

Reply:
Hello @waynechen55,

1) as I already explained, we use test modules from `tests/console/` and those are running not only on virtualization but elsewhere. I cannot include a library from outside of `lib/` there. So yes, not includable means not usable.

2) Yes, in some test modules we use `virt_feature_test_base` but in some it's not possible. Also I had to add there some conditions so it will work for the QAM scenario. You expect to have one guest, we have sometimes none, sometimes 10. You also destroy and recreate your guest after every test module - we don't do so.

3) The bash scripts are hidden for me and are as such very bad practice. There is hardcoded stuff inside and I am just not willing to use them at all. I'm sorry, but I find it obsolete.

If it's OK for you, I'd like to merge this today.

Thank you for your imputs.

Reply:
1. Actually I do not care too much about whether you are going to merge today or tomorrow. I just worry about there will be too many duplicates with regard to logs collecting, for example, there will be no need to use c by us from my perspective. So if you call this in  virt_feature_test_base, you'd better add some conditions.
2. The script used in virt_utils::collect_host_and_guest_logs is standalone one, you can use it everywhere also. And it can be extended and enhanced to fulfill more. I think it can collect pretty complete logs from both host and guest. I can see its effectiveness and usefulness. This is also why I consider it can become a universal logs collecting tool. As a user you  do not need to worry about what is inside, you can just use its exposed interface which provides flexibility instead of "bad practice" and "hardcoded stuff". On the contrary, collect_virt_system_logs looks like that. You of course have freedom to choose what you want to use, but please keep in mind that it should not be at the cost of redundant effort, not-well-organized code and short-sighted way of doing things.
3. As far as virt_feature_test_base is concerned, there is some cleanup actions at the beginning, but just restart guests. And it does not restrict how many number of guests you are going to use. I understand you may have other considerations to make decision of use it or not. It does not matter. 
4. I still insist on using a universal interface to collect logs especially in virt_feature_test_base module if it is going to be used more and more in the future. So please also keep in mind that there might be something can be done to improved the situation instead of just developing another subroutines to collect another type of logs. Hope you can understand and our short discussion here can serve a chance which raise awareness of virt_utils::collect_host_and_guest_logs. It is not that bad as you said.
I currently do not have further comments, just want to remind you that please make sure your change will not break any existing functionality which will be appreciated. It is not my call whether this is going to be merged or not. I just think virt_utils::collect_host_and_guest_logs surpasses virt_utils::collect_host_and_guest_logs. Good Night.

Reply:
1. OK, good to know.
2. Yes, there will be duplicates. We cannot use your function - we need to create usable duplicate.
3. Those cleanup actions break our setup.
4. Feel free to exclude `collect_virt_system_logs` in your scenarios.

IMO Resolved.

Reply:
> @guoxuguang Do you please have specific suggestions of what should I possibly improve?
imho, no just only your part to improve (virt) logs collection. We'd better together to do this(virt_auto and QAM).
So far, we just had created virt_utils::collect_host_and_guest_logs depends on virt_auto test logs collection requirement. 
it do not work on QAM environment very well, which is normal here. 
Moreover, we' d better to use with the same logs collection API (e.g virt_utils::collect_host_and_guest_logs )together(Facilitate future public fun maintenance ). 
For example, use with "supportconfig" utility and so on. 
For now, refer to this PR, we'd better to figure out the same thing depend on virt_auto and QAM logs collection requirements.
We'd create a new poo# (process ticket) to do this together.  @pdostal Hope that help to you.